### PR TITLE
xds/resolver: support inline RDS resource from LDS response

### PR DESCRIPTION
**Pull request by menghanl**
Needed by direct path.

### CODE REVIEW

File: xds/internal/resolver/watch_service.go [hunk 1]
```diff
@@ -110,6 +110,25 @@ func (w *serviceUpdateWatcher) handleLDSResp(update xdsclient.ListenerUpdate, er
 		httpFilterConfig:  update.HTTPFilters,
 	}
 
+	if update.RouteConfigName == "" {
+		// RDS name from update is empty string, the LDS resp received has an
+		// inline RDS resource.
+
+		// If there was an RDS watch, cancel it.
+		if w.rdsName != "" {
```
Comment (dfawley) [inline xds/internal/resolver/watch_service.go:118]: This outer `if` seems unnecessary.

File: xds/internal/resolver/watch_service.go [hunk 2]
```diff
@@ -110,6 +110,25 @@ func (w *serviceUpdateWatcher) handleLDSResp(update xdsclient.ListenerUpdate, er
...
+			}
+			w.rdsCancel = nil
+		}
+
+		// Handle the inline RDS update as if it's from an RDS watch.
+		w.updateVirtualHostsFromRDS(update.InlineRouteConfig)
+		return
+	}
+
+	// RDS name from update is not an empty string, start RDS watch.
```
Comment (dfawley) [inline xds/internal/resolver/watch_service.go:131]: This seems like it's documenting the `if` immediately below, but it isn't.

**Review by dfawley — APPROVED** [review]

File: xds/internal/resolver/watch_service.go (hunk 1 above)
Comment (menghanl) [reply xds/internal/resolver/watch_service.go]: Right

File: xds/internal/resolver/watch_service.go (hunk 2 above)
Comment (menghanl) [reply xds/internal/resolver/watch_service.go]: Modified (a little bit).

**Review by dfawley — APPROVED** [review]

File: xds/internal/client/client.go [hunk 3]
```diff
@@ -195,7 +195,15 @@ type UpdateMetadata struct {
...
 	// RouteConfigName is the route configuration name corresponding to the
 	// target which is being watched through LDS.
+	//
+	// Only one of RouteConfigName and InlineRouteConfig is set.
 	RouteConfigName string
+	// InlineRouteConfig is the inline route configuration (RDS response)
+	// returned inside LDS.
+	//
+	// Only one of RouteConfigName and InlineRouteConfig is set.
+	InlineRouteConfig RouteConfigUpdate
```
Comment (easwars) [inline xds/internal/client/client.go:205]: Is it required/necessary to be able to differentiate between an empty inline `RouteConfigUpdate` and a non-existent one?

Like, should we reject the resource if it does not contain both `RouteConfigName` and inline `RouteConfigUpdate`?

So, should we make this a pointer?

Comment (menghanl) [reply xds/internal/client/client.go]: Made this a pointer.

I was check if `RDS-name` is an empty string. But I think an explicit check for pointer is better.

File: xds/internal/resolver/watch_service.go [hunk 4]
```diff
@@ -110,6 +110,24 @@ func (w *serviceUpdateWatcher) handleLDSResp(update xdsclient.ListenerUpdate, er
 		httpFilterConfig:  update.HTTPFilters,
 	}
 
+	if update.InlineRouteConfig != nil {
+		// The LDS resp received has an inline RDS resource.
```
Comment (easwars) [inline xds/internal/resolver/watch_service.go:114]: Nit: maybe get rid of this comment?

**Review by easwars — APPROVED** [review]

File: xds/internal/resolver/watch_service.go (hunk 4 above)
Comment (menghanl) [reply xds/internal/resolver/watch_service.go]: Done

**Pull request merged**